[Ready] Added animations to cryo/sleepers opening/closing

Razharas opened the pull request:
[Changelogs]: #

:cl: Razharas
add: Animation to cryo and sleepers opening/closing
tweak: Changed a bit of internal handling of visuals of cryo, no more updating icon each time anything is clicked in the ui
imageadd: Sprite for the cryo and sleeper animations

/:cl:

[why]: # Better visuals is better immulsions so yeah
![image](https://i.imgur.com/Ju4moeW.gif)

Razharas committed:
Added animations to cryo/sleepers opening/closing

Also changed a bit of internal handling of icons of cryo
Images -> mutable appearances
No more updating icons every click on anything in the ui

nicbn commented:
Video pls

nicbn commented on code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:
```diff
@@ -89,36 +89,41 @@
 
 		if(ismonkey(occupant)) // Monkey
-			occupant_overlay = image(CRYOMOBS, "monkey")
+			occupant_overlay = mutable_appearance(CRYOMOBS, "monkey")
```
These need to be `image` instead of `mutable_appearances` so that `dir` can be independent of parent. Please fix this and add a comment so that no one will do this again.

Razharas replied:
Currently it follows dir of the cryo even if its image

nicbn replied:
Wrong. Line 123 (or 128): `occupant_overlay.dir = SOUTH`

Razharas replied:
Disregard this i was on the old version of byond so it did follow the locations dir even if it was set manually

nicbn replied:
You removed all `occupant_overlay =...` lines now, you need to replace em

Razharas replied:
Im not sure how did that happen, they were present in the dme

nicbn replied:
Please add a comment in the code so that no one will try to replace `image` again later on

nicbn commented on code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:
```diff
@@ -89,36 +89,41 @@
 		add_overlay("pod-panel")
 
-	if(state_open)
+	if(state_open && (icon_state != "pod-open"))
```
Do not test `icon_state`, add some control var instead.

Razharas replied:
Why wouldnt i check icon state in a proc designated specifically to managing icon states in an environment where we do not track icon states in any specific way besides its factual value?

nicbn replied:
It's a bad practice, icon_states shouldn't be used for logic (even if it's icon logic), only for output.

Razharas replied:
Adding a control var for something which should be handled naturally should we not opted for universal cut and rebuild proc on every occasion is as bad of a cancer as switching machinery on or off based on icon_state

nicbn replied:
Adding a var for logic makes code more maintainable dude, if you check for `icon_state` it'll be painful and error-prone for someone to change the default open `icon_state`

Razharas replied:
Adding a var that traces a very specific transition state from on to off is just as bad
At best its an argument to update_icons() with 3 states: swtiching_to_open, switching_to_close, non_switching_update
But that would need some additional proc call chain changes since machines all use generic open and close procs and they call a generic update_icons there

nicbn commented:
Btw you shouldn't add anything between `[Changelings]: #` and `:cl:`

nicbn commented:
Btw I feel like the cryo animation needs a little more frames

Razharas committed:
Back to images

Razharas commented:
Im coderspriting so adding more than 2 frames is too hard

Razharas committed:
Actually fixes images

nicbn commented:
You're using each frame as a separate icon state and updating the frame code-wise, right?
You should use flick - that way, you can have an animation icon instead that replaces the old icon state and restores the old icon state when it ends.

Razharas commented:
Is there a way to do a time-reverse flick animation?

nicbn commented:
Just duplicate the animation icon (we do this with all reversed animation, like airlocks etc)

Nabski89 commented:
This makes it a pain to try to do custom icons. Please do the variable.

You should be able to set the icon as a non looping "movie" with the end sprite as it in the final location rather than having an intermediate sprite.

Razharas commented:
At the absolute worse ill introduce an argument for an update_icons() proc
Having an actual permanent var which is basically a stack argument placed on the heap coz our call chain is special olympics kind of special is not a solution ill agree to

SpaceManiac commented:
Admin varediting is a supported usecase.

Avoid predicating logic on `icon_state ==`, it makes the code harder to change down the line.

Consider using the `animate()` proc to play the animation rather than setting `icon_state` on a timer.

Razharas commented:
Its not about performance its aabout code consistency
ALso this pr is in wait for lummox to fix his new icon editor in dm coz neither i not wj can make it work like intended

nicbn commented:
There's no code consistency on it, every other use case of it uses (or should use) var instead of icon_state.

SpaceManiac commented:
WRT the code consistency, I discussed this with Razharas in IRC and recommend using animate() or flick() in open_machine()/close_machine() rather than adding animation logic to update_icon()

ExcessiveUseOfCobblestone commented:
Update?

SpaceManiac commented:
Your PR has been marked Stale because it has been marked Work In Progress or Changes Requested for an extended time with no updates. A maintainer may close the PR if you do not update it. If you want to continue work on the PR after that, ping any maintainer to have it reopened.

(If you're waiting on WJ, it's up to you whether to wait before reopening or submit this and then PR WJ's sprites when they're done)

Razharas committed:
Use of flick and animation sprites

Changed the way animations are done to use flick proc instead of animate
proc
Changed the way animations of cryo cells are organazed in a dmi, now
separate icon_state for opening and closing

ExcessiveUseOfCobblestone commented:
when ur ready remove WIP from title and the bot should handle the rest :3

ExcessiveUseOfCobblestone commented:
How does this react to sleeper subtypes?

ExcessiveUseOfCobblestone reviewed (commented):
^^

Razharas commented:
Sleepers have subtypes?

Razharas committed:
Support for resprites of the cryo subtypes

Support for resprites of the cryo subtypes

ExcessiveUseOfCobblestone commented:
Current iteration won't support varediting

Razharas commented:
If varediting is now an actual use case for development than close this pr, i wont support varediting being use case

ExcessiveUseOfCobblestone commented:
Someone else can modularize it then hehe
